### PR TITLE
Add VPC and IAM resources to request_quota_increase + code gen script

### PR DESCRIPTION
**Pull request by marinalimeira**
**THIS PR IS BACKWARDS INCOMPATIBLE** the existing keys, `nat_gateway` and `nacl_rules`, were changed.

This is an improvement of #38.

Changes:
- New script to generate code to generate the necessary code to add a new resource
- No need to use the 'Service Code' to lookup for resources, using [aws_servicequotas_service_quota](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/servicequotas_service_quota).
- `nat_gateway` renamed to `vpc_nat_gateways_per_availability_zone` to follow new pattern
- `nacl_rules` renamed to `vpc_rules_per_network_acl` to follow new pattern

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

File: modules/request-quota-increase/generate_code.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# ruby generate_code.rb
+#################################################################
+
+require 'json'
+
+SERVICE = 'iam'
+
+# Some services are not available in all regions. For example, 'route53' and 'iam'
+# are only available in 'us-east-1'
+REGION = 'us-east-1'
```
Comment (brikis98) [inline modules/request-quota-increase/generate_code.rb:16]: Perhaps allow these to be passed in as CLI args?

```
ruby generate_code.rb <SERVICE> <REGION>
```

File: modules/request-quota-increase/generate_code.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
+#################################################################
+# REQUEST QUOTA INCREASE CODE GENERATOR
+# This script generates the code for add more resources into the
+# module. It only generates the code for the resources that
+# 'Adjustable' is true.
```
Comment (brikis98) [inline modules/request-quota-increase/generate_code.rb:5]: This needs a bit of context first: i.e., that AWS requires you to pass in special codes to request quotas, that there is no central place where these are all defined, that there are many of these, we've included some in the `request-quota-increase` module already, but to add the others, this script can be used on a service-by-service basis.

File: modules/request-quota-increase/README.md [hunk 3]
```diff
@@ -114,3 +99,18 @@ new_quotas = {
 When you run `terraform destroy` on this module, it does not affect your current quotas or your
 existing quota requests. In other words, you don't have to worry about quotas being reset to old
 values; once they have been increased, they stay that way!
+
+## Contributing
```
Comment (brikis98) [inline modules/request-quota-increase/README.md:103]: Perhaps this section should have a title like: "How do I request quota increases for services this module doesn't support yet?"

File: modules/request-quota-increase/README.md [hunk 4]
```diff
@@ -114,3 +99,18 @@ new_quotas = {
...
+
+When you want to add a new resource, you can check the available services with
```
Comment (brikis98) [inline modules/request-quota-increase/README.md:105]: This sections needs a bit of context first: i.e., that AWS requires you to pass in special codes to request quotas, that there is no central place where these are all defined, that there are many of these, we've included the values for the IAM and VPC services in the `request-quota-increase` module already, but to add other services, we've created a script...

File: modules/request-quota-increase/main.tf [hunk 5]
```diff
@@ -3,24 +3,149 @@ terraform {
...
-      quota_code   = "L-2AEEBF1A"
+
+  resources_code = {
+    for key, value in var.resources_to_increase :
+    key => {
+      quota_code   = data.aws_servicequotas_service_quota.by_quota_name[key]["quota_code"]
+      service_code = local.resource_names[key].service_code
+      value        = value
+    } if !(data.aws_region.current.name != "us-east-1" && substr(key, 0, 3) == "iam")
+    # It will ignore the resource if it's IAM and it is not on us-east-1 region
```
Comment (brikis98) [inline modules/request-quota-increase/main.tf:31]: NIT: Update comment to explain that IAM _only_ works in us-east-1, which is why you need to skip it in other regions.

File: modules/request-quota-increase/main.tf [hunk 6]
```diff
@@ -3,24 +3,149 @@ terraform {
...
+  }
+
+  # PRs to add more of these mappings are very welcome.
```
Comment (brikis98) [inline modules/request-quota-increase/main.tf:34]: NIT: update comment to say this was generated using the generate_code.rb script for a subset of AWS resources. PRs to add others are welcome.

File: modules/request-quota-increase/variables.tf [hunk 7]
```diff
@@ -1,10 +1,10 @@
 variable "resources_to_increase" {
-  description = "A map where the key is the resource and the value is the desired quota. The only resources supported at the moment are 'nacl_rules' and 'nat_gateway'. You can also use the `aws_servicequotas_service_quota` resource directly, there are instructions on how to find the Service Code and Quota Code on the README!"
+  description = "A map where the key is the resource and the value is the desired quota. The only services supported at the moment are VPC and IAM. You can also use the `aws_servicequotas_service_quota` resource directly!"
```
Comment (brikis98) [inline modules/request-quota-increase/variables.tf:2]: We should tell users where to look up the available values. 

Also, what does "You can also use the aws_servicequotas_service_quota resource directly" mean?

**Review by brikis98 — COMMENTED** [review]
Nice!

File: modules/request-quota-increase/generate_code.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
+#################################################################
+# REQUEST QUOTA INCREASE CODE GENERATOR
```
[the remark this answers is not in the document]
Comment (yorinasub17) [reply modules/request-quota-increase/generate_code.rb:2]: Not for this PR, but should we standardize the languages we use? I'm concerned that we now have ruby in the mix of languages we need and I don't really have a ruby environment setup.

[the remark this answers is not in the document]
Comment (brikis98) [reply modules/request-quota-increase/generate_code.rb:2]: Good question. There are a bunch of pros/cons to standardization... Might need a deeper discussion, either async (e.g., in Notion) or at GruntCon.

File: modules/request-quota-increase/variables.tf (hunk 7 above)
Comment (marinalimeira) [reply modules/request-quota-increase/variables.tf:2]: > Also, what does "You can also use the aws_servicequotas_service_quota resource directly" mean?
That if there is not available not the module, they can use just the resource.

File: modules/request-quota-increase/generate_code.rb (hunk 1 above)
Comment (marinalimeira) [reply modules/request-quota-increase/generate_code.rb:16]: Good idea!

File: modules/request-quota-increase/generate_code.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+def to_camel_case str 
+	"#{SERVICE.capitalize}#{str.gsub('-', ' ').split(' ').collect(&:capitalize).join}"
+end
+
+def to_snake_case str
+	"#{SERVICE.downcase}_#{str.downcase.gsub(" ", "_").gsub('-', '_')}"
+end
+
+quotas = JSON.parse(`aws service-quotas list-service-quotas --service-code #{SERVICE} --region #{REGION} | jq '[ .Quotas[] | select(.Adjustable) ]'`)
```
[the remark this answers is not in the document]
Comment (marinalimeira) [reply modules/request-quota-increase/generate_code.rb]: Good catch! Removed `jq`. I updated to use `filter` from ruby.

**Comment by rhoboat** [comment]
@marinalimeira This is an old PR, but I like the changes. Can we fix the conflicts, merge this up and take care of NITs later?

**Review by rhoboat — DISMISSED** [review]
This is an old PR, but I like the changes. Can we fix the conflicts, merge this up and take care of NITs later?

File: modules/request-quota-increase/generate_code.rb (hunk 2 above)
Comment (marinalimeira) [reply modules/request-quota-increase/generate_code.rb]: Fixed! 4b3b8e6

File: modules/request-quota-increase/variables.tf (hunk 7 above)
Comment (marinalimeira) [reply modules/request-quota-increase/variables.tf]: Updated! 1e19c1e

**Review by marinalimeira — COMMENTED** [review]
@rhoboat thank you for the heads up about this stale PR! It has no conflicts with `main` anumore and I update the code with the suggestions.

**Comment by marinalimeira** [comment]
Hm, the tests are failing, I just locked the region to us-east-1...

**Comment by marinalimeira** [comment]
It is still failing but it works locally for me... I will investigate this on Monday.

**Comment by marinalimeira** [comment]
Sooooo... I finally took the proper time to take a look at this. I was running the tests in the wrong branch, that's why it was working locally...

Well, there was a tiny typo in the test, it's fixed now! cb40b56 Tests will hopefully pass in this run and we can merge it!

@rhoboat can you take a final look? :)

**Review by rhoboat — COMMENTED** [review]
Now that we're writing patches for backward incompatible updates, I have some requests. 😅 I'm sorry, this looks like a lot:

- Can we first merge: [PR to update the upgrade tests](https://github.com/gruntwork-io/terraform-aws-utilities/pull/67/files)
- Then could we add an upgrade test for the example `request-quota-increase`?
    - We should see that test fail.
    - Then we should add a patch in the patches directory.
    - That patch should be applied in the upgrade test, automatically.
    - The upgrade test should then pass.

Then, it would be helpful to put in the PR description:
- A draft of the release notes
- A migration guide (the steps a user has to run, pre-Patcher)

**Comment by MoonMoon1919** [comment]
@marinalimeira Closing this as part of a PR clean-up initiative. Feel free to re-open if you continue to plan on working on this!

**Pull request closed without merging**